Migrate JupterLab from 2.x to 3.x

NikitaGandhi opened the pull request:
- JupyterLab 3.x comes with many inbuilt features - https://blog.jupyter.org/jupyterlab-3-0-is-out-4f58385e25bb
- JupyterLab releases prebuilt extensions and available to PyPI - reduce step to install JupyterLab extensions and rebuild  JupyterLab

Commit:
remove comments

Commit:
JupyterLab migration from 2.x to 3.x

NikitaGandhi commented:
This PR requires extended testing before updating sandbox environment and not ready to merge :)

Commit:
enable JupyterLab extentions

Commit:
use stable release for jupyterlab-widgets

Commit:
cleanup

Commit:
update libs

Commit:
enable jupyter jupyterlab_iframe serverextension

NikitaGandhi committed:
pinning dask related libs

Commit:
updated contraints

Commit:
update dask version to 2021.7.1

NikitaGandhi committed:
update odc libs version

alexgleith commented on docker/constraints-odc.txt:
```diff
@@ -1,16 +1,16 @@
+odc_dtools>=0.2.1.dev
+odc_stats>=0.2.1.dev
+odc-apps-dc-tools>=0.2.1.dev
+odc-apps-cloud>=0.2.1.dev
```
These libraries are all in flux right now, we may want to wait for them to be in PyPI and pull from there... Or deal with that change later!

alexgleith reviewed (approved):
An enormous list of library changes. Comprehensive and awesome, Nikita!

omad commented:
Is it possible to roll this out in parallel with the existing version 2.x?

It'd be great to be able to test without potentially breaking things for everyone.

alexgleith commented:
Do you mean we maintain two different image builds, @omad? If so, then I'm not sure that it's a good idea to maintain two environments, as we struggle enough maintaining one!

We have got this image deployed in parallel, though, so they can both be tested.

omad commented:
> Do you mean we maintain two different image builds, @omad? If so, then I'm not sure that it's a good idea to maintain two environments, as we struggle enough maintaining one!

Nhooo, definitely do not want that.

> We have got this image deployed in parallel, though, so they can both be tested.

Awesome, that's what I was hoping for, and is what's needed for testing and approving this kind of change. :D

robbibt commented:
Hey @NikitaGandhi, a silly question: where can I actually test this? Is it deployed on a Sandbox somewhere?

Commit:
install JupyterLab extensions to display system information - memory and cpu usage

Commit:
install jupyter top bar jupyterlab-logout extension

NikitaGandhi committed:
upgrade bokeh, install pyviz_comms - Jupyterlab pyviz extension

NikitaGandhi committed:
downgrade nbconvert

NikitaGandhi committed:
commented out panel and nbresuse

NikitaGandhi committed:
add jupyter-server and panel libs

NikitaGandhi committed:
pin dask and distributed to 2.30.0 due to dask-gateway dependency

NikitaGandhi committed:
add cognitojwt

cbur24 commented:
Something weird I'm noticing...when I first open a notebook and use "restart kernel and run all cells", some of the cells in the middle of the notebook are skipped (not evaluated). If I clear the kernel and do it again, it works fine.

![image](https://user-images.githubusercontent.com/26324916/132788817-ff5f86c2-7e36-4809-b554-d5203160fa74.png)

cbur24 commented:
we still have this hacky memory counter in the bottom left of the window, I think we should remove it given we now have the graphic in the top-left corner

![image](https://user-images.githubusercontent.com/26324916/132788936-addd2fc6-756a-4bcd-9ce5-bf7980ec708f.png)

cedricump commented:
Can we ensure that the French language pack is install by default. so user can readily access the language through JupyterLab settings without pip installing it? the latest version is 100% translated. Tested on dev through pip install and it is working.

NikitaGandhi committed:
install odc packages from pypi

Commit:
fix odc-aio installation

Commit:
install odc-stac from pypi

NikitaGandhi committed:
revised odc requirements

NikitaGandhi committed:
remove odc-stac from the list

NikitaGandhi committed:
try disabling jupyter-resource-usage

NikitaGandhi committed:
try disabling jupyter-resource-usage - take2

Commit:
disable jupyter_resource_usage - take3

NikitaGandhi committed:
revert previous chage - keep jupyter_resource_usage needed for jupyterlab-system-monitor

NikitaGandhi committed:
install jupyterlab-theme-toggle

NikitaGandhi commented:
> we still have this hacky memory counter in the bottom left of the window, I think we should remove it given we now have the graphic in the top-left corner
> 
> ![image](https://user-images.githubusercontent.com/26324916/132788936-addd2fc6-756a-4bcd-9ce5-bf7980ec708f.png)

- Updated top-bar extention to include `CPU usage`, `Memory usage`, `Logout button` and `theme-toggle button` and removed `jupyterlab-topbar-text` ("Hi there" text on topbar):

![image](https://user-images.githubusercontent.com/2030490/133174517-a8eb6930-166a-455b-8873-d3011277e85a.png)

- Not able to remove/disable memory usage metrics from bottom as this settings comes from `jupyter-resource-usage` extension as default and also required by top-bar `jupyterlab-system-monitor` extension. But the bottom status-bar is enable/disable from menu `view -> Show status bar` by user as an option.

Kirill888 commented on docker/constraints-odc.txt:
```diff
@@ -1,16 +1,21 @@
\ No newline at end of file
+odc-aio==0.2.1.dev2476
+odc-cloud==0.2.1.dev2644
+odc-index==0.2.1.dev2420
```
no such thing anymore, it was renamed into `odc-stac`

Kirill888 commented on docker/constraints-odc.txt:
```diff
@@ -1,16 +1,21 @@
+odc-aio==0.2.1.dev2476
+odc-cloud==0.2.1.dev2644
+odc-index==0.2.1.dev2420
+odc-aws==0.2.1.dev2476
+odc-ppt==0.2.1.dev2400
+odc-dtools==0.2.1.dev2400
```
no such thing anymore, those got merged/deleted `odc-cloud`.

NikitaGandhi committed:
revise odc libs as per odc-tools docs - pulling from PyPi

Kirill888 commented on docker/constraints-odc.txt:
```diff
@@ -1,21 +1,10 @@
-odc-geom==0.2.0a0
 odc-io==0.2.0a0
-odc_stats==0.2.0a2
+odc-stac==0.2.0a4
```
can we please bump this one:
```suggestion
odc-stac==0.2.0a5
```

NikitaGandhi committed:
upgrade numpy to latest 1.21.2

NikitaGandhi committed:
updated odc-stac and odc-stats versions

NikitaGandhi committed:
updated constraint file

NikitaGandhi committed:
install french language pack

NikitaGandhi commented:
> Can we ensure that the French language pack is install by default. so user can readily access the language through JupyterLab settings without pip installing it? the latest version is 100% translated. Tested on dev through pip install and it is working.

added

NikitaGandhi committed:
odc-tools packages is behind so pulling it from dea-packages repo

NikitaGandhi committed:
fix constraints

NikitaGandhi committed:
install jupyter-resource-usage - require for resource usage info

NikitaGandhi committed:
update jupyter server extention - getting no module jupyter_resource_usage in singleserver app

mmochan reviewed (approved):
LGTM!
